[ONNX] Stablize eraseTupleConstruct

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76055
* #76002
* #75921
* __->__ #75920

Previous implementation is prone to segfault, as it is inserting
new outputs to block while iterating over block->outputs().

BowenBao committed:
[ONNX] Stablize eraseTupleConstruct

Previous implementation is prone to segfault, as it is inserting
new outputs to block while iterating over block->outputs().

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Stablize eraseTupleConstruct"


Previous implementation is prone to segfault, as it is inserting
new outputs to block while iterating over block->outputs().

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Stablize eraseTupleConstruct"


Previous implementation is prone to segfault, as it is inserting
new outputs to block while iterating over block->outputs().

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Stablize eraseTupleConstruct"


Previous implementation is prone to segfault, as it is inserting
new outputs to block while iterating over block->outputs().

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Stablize eraseTupleConstruct"


Previous implementation is prone to segfault, as it is inserting
new outputs to block while iterating over block->outputs().

[ghstack-poisoned]

garymm commented on torch/csrc/jit/passes/onnx/peephole.cpp:
```diff
@@ -763,23 +763,24 @@ static void fuseListConstructListUnpack(Block* b) {
+      new_block_outputs.emplace_back(output);
     }
   }
+  block->removeAllOutputs();
```
not sure if this is a premature optimization, but maybe keep a bool like:
`found_tuple_construct` and only remove and register new outputs if it's true?

Up to you if you think this is a good idea.

BowenBao replied:
updated.

BowenBao committed:
Update on "[ONNX] Stablize eraseTupleConstruct"


Previous implementation is prone to segfault, as it is inserting
new outputs to block while iterating over block->outputs().

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge this
